### PR TITLE
CUMULUS 226 additional: kinesis rules can share the same event source mapping

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS 226: Support "subscription"-type rules in Cumulus Rules API](https://bugs.earthdata.nasa.gov/browse/CUMULUS-226)

## Test Plan
Tested with cumulus-dashboard and added unit test

### CODE REVIEW

File: packages/api/models/rules.js [hunk 1]
```diff
@@ -148,6 +152,23 @@ class Rule extends Manager {
...
    */
   async addKinesisEventSource(item) {
+    // use the existing event source mapping if it already exists
+    const listParams = {
+      EventSourceArn: item.rule.value,
+      FunctionName: process.env.kinesisConsumer
+    };
+    const listData = await aws.lambda(listParams).listEventSourceMappings().promise();
+    if (listData.EventSourceMappings && listData.EventSourceMappings.length > 0) {
+      const maps = listData.EventSourceMappings;
```
Comment (abarciauskas-bgse) [inline packages/api/models/rules.js]: I think we can use `Array.prototype.find` to simplify this loop, something like:
```javascript
if (listData.EventSourceMappings) {
  const eventSourceMappingExists = listData.EventSourceMappings.find((eventSourceMapping) => {
    return eventSourceMapping.EventSourceArn === item.rule.value;
  });
  if (eventSourceMappingExists) {
    item.rule.arn = maps[i].UUID;
    return item;
  }
}
```

I didn't test this so if it doesn't work for some reason this looks fine as-is

File: packages/api/models/rules.js [hunk 2]
```diff
@@ -181,12 +202,41 @@ class Rule extends Manager {
    * @returns {Promise} the response from event source delete
    */
   async deleteKinesisEventSource(item) {
+    if (await this.isEventSourceMappingShared(item)) {
```
Comment (abarciauskas-bgse) [inline packages/api/models/rules.js]: Could we add a test for this, e.g. `test('it does not delete event source mapping if it exists for other rules')`?

File: packages/api/models/rules.js [hunk 3]
```diff
@@ -181,12 +202,41 @@ class Rule extends Manager {
...
+      return {};
```
Comment (abarciauskas-bgse) [inline packages/api/models/rules.js]: Any reason to return an empty object here? I think it makes more sense to return undefined (e.g. just `return;`)

File: packages/api/tests/test-rules-model.js [hunk 4]
```diff
@@ -117,4 +119,29 @@ test('update a kinesis type rule value, resulting in new arn', async (t) => {
...
+
+  t.not(newRule.name, rule.name);
+  t.is(newRule.rule.value, rule.rule.value);
+  t.false(newRule.rule.arn === undefined);
+  // same event source mapping
+  t.is(newRule.rule.arn, rule.rule.arn);
+
+  await rules.delete(rule);
+  await rules.delete(newRule);
 });
```
Comment (abarciauskas-bgse) [inline packages/api/tests/test-rules-model.js]: Not necessarily for this PR but I think we should test our expectations on creating / updating / deleting the lambda event source mappings, I'll add these as todo specs in a separate PR

Comment (jennyhliu) [reply packages/api/tests/test-rules-model.js]: I think these tests covered the event source mapping. e.g.
test('update a kinesis type rule value, resulting in new arn')

File: packages/api/models/rules.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/models/rules.js]: Ok.  Updated. I just didn't want it to return failure.

File: packages/api/tests/test-rules-model.js (hunk 4 above)
Comment (abarciauskas-bgse) [reply packages/api/tests/test-rules-model.js]: I see what you are saying but I think we could use a lot more tests here - this is what I am thinking: https://github.com/cumulus-nasa/cumulus/compare/message-updates...abarciauskas-bgse_kinesis-consumer-event-source-mappings-tests - but I wouldn't worry about adding those in this PR!

File: packages/api/models/rules.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/models/rules.js]: Actually, eslint will give me error for next return statement if "return;" is used here.
"Async request expected no return value"

File: packages/api/models/rules.js (hunk 2 above)
Comment (jennyhliu) [reply packages/api/models/rules.js]: done

File: packages/api/models/rules.js (hunk 1 above)
Comment (jennyhliu) [reply packages/api/models/rules.js]: your find method works!

**Comment by jennyhliu** [comment]
I have addressed your comments, please review and approve.  Thanks!@abarciauskas-bgse

**Pull request merged**